src/attach-shadow.js

```diff
@@ -237,25 +293,141 @@ ShadyRoot.prototype._updateChildNodes = function(container, children) {
     for (let j=s.index, n; j < s.index + s.addedCount; j++) {
       n = children[j];
       insertBefore.call(container, n, next);
-      // TODO(sorvell): is this splice strictly needed?
       composed.splice(j, 0, n);
     }
   }
 }
 
-ShadyRoot.prototype.getInsertionPointTag = function() {
-  return this._distributor.insertionPointTag;
+/**
+ * Adds the given slots. Slots are maintained in an dom-ordered list.
+ * In addition a map of name to slot is updated.
+ */
+ShadyRoot.prototype._addSlots = function(slots) {
+  let slotNamesToSort;
+  for (let i=0; i < slots.length; i++) {
+    let slot = slots[i];
+    // ensure insertionPoints's and their parents have logical dom info.
+    // save logical tree info
+    // a. for shadyRoot
+    // b. for insertion points (fallback)
+    // c. for parents of insertion points
+    slot.__shady = slot.__shady || {};
+    recordChildNodes(slot);
+    recordChildNodes(slot.parentNode);
+    let name = this._nameForSlot(slot);
+    if (this._slotMap && this._slotMap[name]) {
+      slotNamesToSort = slotNamesToSort || {};
+      slotNamesToSort[name] = true;
+    }
+    this._slotList.push(slot);
+  }
+  if (slotNamesToSort) {
+    for (let n in slotNamesToSort) {
+      let slotsForName = this._extractSlotsOfName(n);
+      let sorted = this._sortSlots(slotsForName);
+      this._slotList.push(...sorted);
+    }
+  }
+  this._updateSlotMap();
+}
+
+ShadyRoot.prototype._nameForSlot = function(slot) {
+  return slot['name'] || slot.getAttribute('name') || CATCHALL_NAME;
+}
+
+ShadyRoot.prototype._extractSlotsOfName = function(name) {
+  let slots;
+  for (let i=0; i < this._slotList.length; i++) {
+    let slot = this._slotList[i];
+    let n = this._nameForSlot(slot);
+    if (n == name) {
+      this._slotList.splice(i, 1);
+      i--;
+      slots = slots || [];
+      slots.push(slot);
+    }
+  }
+  return slots;
+}
+
+/**
+ * Slots are kept in an ordered list. Here they are sorted by dom-tree order.
+ */
+ShadyRoot.prototype._sortSlots = function(slots) {
+  return slots.sort((a, b) => {
+    let listA = ancestorList(a);
+    let listB = ancestorList(b);
+    for (var i=0; i < listA.length; i++) {
+      let nA = listA[i];
+      let nB = listB[i];
+      if (nA !== nB) {
+        let c$ = Array.from(nA.parentNode.childNodes);
+        return c$.indexOf(nA) - c$.indexOf(nB);
+      }
+    }
+  });
+}
+
+function ancestorList(node) {
+  let ancestors = [];
+  do {
+    ancestors.unshift(node);
+  } while ((node = node.parentNode));
+  return ancestors;
+}
+
+function contains(container, node) {
+  while (node) {
+    if (node == container) {
+      return true;
+    }
+    node = node.parentNode;
+  }
+}
+
+/**
+ * Given a `container` element, removes any contained slot elements,
```

Comment:
Comment improvement:
> removes any contained slot elements _from the `_slotList`...

I read it as "removes them from the dom" and was like wat